opt: build all partial index predicate expressions in TableMeta

mgartner opened the pull request:
#### opt: build all partial index predicate expressions in TableMeta

Previously, partial index predicates were only built and added to
`TableMeta.PartialIndexPredicates` if the scope of the scan in
`Builder.buildScan` included all ordinary table columns. This behavior
prevented errors when building partial index predicate expressions for
foriegn key check scans. These scans do not include all columns, so when
building a predicate expression that referenced a table column not
included, optbuilder would err with "column does not exist".

This commit changes this behavior so that partial index predicates are
always built and added to `TableMeta.PartialIndexPredicates`. In order
to do this, `Builder.addPartialIndexPredicatesForTable` creates its own
table scope. It also constructs a new scan rather than relying on the
already-built scan. Constructing a scan is required for considering
logical properties while normalizing partial index predicates.

See discussion at https://github.com/cockroachdb/cockroach/pull/57622
for more context on this change.

Release note: None

#### optbuilder: reduce redundant building of arbiter filter expressions

Previously, the arbiter predicate provided in the `WHERE` clause of an
`INSERT ... ON CONFLICT` statement was rebuilt as a `memo.FiltersExpr`
repetitively while the optbuilder determined arbiter indexes. Now, the
expression is built only once, reducing work.

Release note: None

mgartner committed:
opt: build all partial index predicate expressions in TableMeta

Previously, partial index predicates were only built and added to
`TableMeta.PartialIndexPredicates` if the scope of the scan in
`Builder.buildScan` included all ordinary table columns. This behavior
prevented errors when building partial index predicate expressions for
foriegn key check scans. These scans do not include all columns, so when
building a predicate expression that referenced a table column not
included, optbuilder would err with "column does not exist".

This commit changes this behavior so that partial index predicates are
always built and added to `TableMeta.PartialIndexPredicates`. In order
to do this, `Builder.addPartialIndexPredicatesForTable` creates its own
table scope. It also constructs a new scan if the provided scan does not
output all column in the table scope. A scan on the table and its
logical properties are required in order to fully normalize partial
index predicates.

See discussion at https://github.com/cockroachdb/cockroach/pull/57622
for more context on this change.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57707)
<!-- Reviewable:end -->

mgartner committed:
optbuilder: reduce redundant building of arbiter filter expressions

Previously, the arbiter predicate provided in the `WHERE` clause of an
`INSERT ... ON CONFLICT` statement was rebuilt as a `memo.FiltersExpr`
repetitively while the optbuilder determined arbiter indexes. Now, the
expression is built only once, reducing work.

Release note: None

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57707#-:-MO4D-BF00PGY4RwhNj5:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/optbuilder/select.go, line 716 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57707#-MO4CG_25ug_fXJJOsZ9:-MO4CG_25ug_fXJJOsZA:bh9gutq) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d65dd09446bb7276e034cbf7b8ff1e3069fd23f/pkg/sql/opt/optbuilder/select.go#L716)):*
> ```Go
> 	tableScope := b.allocScope()
> 	tableScope.appendOrdinaryColumnsFromTable(tabMeta, &tabMeta.Alias)
> 	tableScope.expr = b.factory.ConstructScan(&memo.ScanPrivate{
> ```

Ah.. constructing the extra Scan is unfortunate.. I guess we could get the FDs (`memo.MakeTableFuncDep`) etc and put them in a `FakeRel` but I don't know if that saves much.

We could pass an optional scan expression to the function, and have this code use it if its OutputCols contains `tableScope.colSet()`. Feels a bit hacky but it will save this construction in most cases.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 9 of 9 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57707#-:-MO6fAx55WjKvD8kbWEW:bg7igkv)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57707)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/opt/optbuilder/select.go, line 716 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57707#-MO4CG_25ug_fXJJOsZ9:-MO9Ts32CPhMMvuUH2tP:bbl3ypn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d65dd09446bb7276e034cbf7b8ff1e3069fd23f/pkg/sql/opt/optbuilder/select.go#L716)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Ah.. constructing the extra Scan is unfortunate.. I guess we could get the FDs (`memo.MakeTableFuncDep`) etc and put them in a `FakeRel` but I don't know if that saves much.

We could pass an optional scan expression to the function, and have this code use it if its OutputCols contains `tableScope.colSet()`. Feels a bit hacky but it will save this construction in most cases.
</blockquote></details>

In the common case wouldn't the `ScanPrivate`s be the same? Therefore`ConstructScan` would return the memoized scan instead of building a new one?


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57707)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/opt/optbuilder/select.go, line 716 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57707#-MO4CG_25ug_fXJJOsZ9:-MO9UXPuEXGG7DpQhDqf:buo1eh2) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d65dd09446bb7276e034cbf7b8ff1e3069fd23f/pkg/sql/opt/optbuilder/select.go#L716)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

In the common case wouldn't the `ScanPrivate`s be the same? Therefore`ConstructScan` would return the memoized scan instead of building a new one?
</blockquote></details>

Nevermind... All the flags would have to be the same...


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57707)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @rytaft)

___
*[pkg/sql/opt/optbuilder/select.go, line 716 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57707#-MO4CG_25ug_fXJJOsZ9:-MOCvRg-BTYthUvSlvt9:b-ptmqm7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d65dd09446bb7276e034cbf7b8ff1e3069fd23f/pkg/sql/opt/optbuilder/select.go#L716)):*
> We could pass an optional scan expression to the function, and have this code use it if its OutputCols contains tableScope.colSet(). Feels a bit hacky but it will save this construction in most cases.

Done. Notice that the size in KB of the memos in the `xform/select` tests are no longer increased by this PR.


<!-- Sent from Reviewable.io -->

mgartner commented:
@RaduBerinde mind taking another look and make sure we're on the same page with this before I merge?

RaduBerinde reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57707#-:-MOIxipcDwkAkQ8E2Ca8:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)



<!-- Sent from Reviewable.io -->

mgartner commented:
TFTRs!

bors r+
